django/chat/responses.py

```diff
@@ -24,44 +26,50 @@
     summarize_long_text_async,
     url_to_text,
 )
-from librarian.models import Document
+from librarian.models import DataSource, Document, Library
 from otto.utils.decorators import permission_required
 
 logger = get_logger(__name__)
 
 
 @permission_required("chat.access_message", objectgetter(Message, "message_id"))
-def otto_response(request, message_id=None):
+def otto_response(request, message_id=None, switch_mode=False, skip_agent=False):
     """
     Stream a response to the user's message. Uses LlamaIndex to manage chat history.
     """
     response_message = Message.objects.get(id=message_id)
     chat = response_message.chat
-    assert chat.user_id == request.user.id
     mode = chat.options.mode
 
     # For costing and logging. Contextvars are accessible anytime during the request
     # including in async functions (i.e. htmx_stream) and Celery tasks.
     bind_contextvars(message_id=message_id, feature=mode)
 
+    agent_enabled = not skip_agent and mode == "chat" and chat.options.chat_agent
+    if agent_enabled:
+        return chat_agent(chat, response_message)
     if mode == "chat":
-        return chat_response(chat, response_message)
+        return chat_response(chat, response_message, switch_mode=switch_mode)
```

Comment:
At present, it will never switch mode into chat mode, but this is ready if we change our minds.